### PR TITLE
Shotgun Shell Bandoliers

### PR DESCRIPTION
**Pull request by Cyaneer**
# What this PR does

Bandoliers!
A belt slot item that stores shells like the quiver, is able to store mixed types (of the same gauge).
Alt-click to take out 5 at a time! Wow!
Three types, one for each faction, as each faction has a different gauge.
Now dyeable! Wowee!

# Why it should be added to the game

- Store shells in your belt slot!
- Grab handfuls at a time!
- No annoying box juggling!
- Flex on bow nerds!

### CODE REVIEW

**Comment by ChewedPencil** [comment]
Exactly what i neeed in my Shotty Life

**Comment by MaintClown** [comment]
dear god please speedmerge

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	. = ..()
+	for(var/k in object_data["stored_shells"])
+		var/v = object_data["stored_shells"][k]
+		stored_shells[text2path(k)] = v
+
+/obj/item/clothing/belt/bandolier/Finalize()
+	. = ..()
+	update_sprite()
+
+/obj/item/clothing/belt/bandolier/proc/get_shell_count()
```
Comment (BurgerLUA) [inline code/_core/obj/item/clothing/belt/bandolier.dm:40]: Use length(stored_shells) for this. it does the same thing here but faster and much better.

**Review by BurgerLUA — CHANGES_REQUESTED** [review]
Seems like a copypaste of quivers, but I can merge for now I guess if everything works.

File: code/_core/obj/item/clothing/belt/bandolier.dm (hunk 1 above)
Comment (Cyaneer) [reply code/_core/obj/item/clothing/belt/bandolier.dm]: Yep, did it. Any particular reason that's used in the quiver dm then?

**Comment by Cyaneer** [comment]
Also yeah, it's more or less just a moderately modified quiver.

**Comment by jjpark-kb** [comment]
I feel there is a way to compact this

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+	)
+
+	rarity = RARITY_UNCOMMON
+	value = 200
+
+	var/list/stored_shells = list()
+	var/shell_count = 0
+
+	var/max_shells = 0
+
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:22]:
```suggestion
/obj/item/clothing/belt/bandolier/clicked_on_by_object(var/mob/caller,var/atom/object,location,control,params)

	if(istype(object,shell_type))
		INTERACT_CHECK
		INTERACT_CHECK_OBJECT
		INTERACT_DELAY(2)
		var/obj/item/bullet_cartridge/S = object
		if(caller.attack_flags & CONTROL_MOD_ALT)
			if(!stored_shells[S.type])
				caller.to_chat(span("warning","There are no shells of that type left in \the [src.name]!"))
				return TRUE
			var/amount_added = S.add_item_count(1)
			if(amount_added)
				stored_shells[S.type] -= amount_added
				if(stored_shells[S.type] <= 0)
					stored_shells -= S.type
				update_sprite()
		else
			var/amount_added = -S.add_item_count(-min(S.item_count_current,max(0,max_shells - length(stored_shells))))
			if(amount_added)
				if(!stored_shells[S.type])
					stored_shells[S.type] = amount_added
				else
					stored_shells[S.type] += amount_added
				update_sprite()
		return TRUE

	if(istype(object,/obj/hud/inventory))
		INTERACT_CHECK
		INTERACT_CHECK_OBJECT
		INTERACT_DELAY(2)
		var/obj/hud/inventory/I = object

		if(!length(stored_shells))
			caller.to_chat(span("warning","There are no shells left!"))
			return TRUE

		var/obj/item/bullet_cartridge/S = pickweight(stored_shells)
		if(caller.attack_flags & CONTROL_MOD_ALT)
			S = new S(get_turf(src))
			var/amount = 0
			if(stored_shells[S.type] < 5)
				amount = stored_shells[S.type]
			else
				amount = 5
			S.item_count_current = amount
			stored_shells[S.type] -= amount
			caller.to_chat(span("notice","You take [amount] shell\s from \the [src.name]. There are [stored_shells[S.type]] shells of that type left."))
			if(stored_shells[S.type] <= 0)
				stored_shells -= S.type
			INITIALIZE(S)
			FINALIZE(S)
			I.add_object(S)
			update_sprite()
			return TRUE
		S = new S(get_turf(src))
		S.item_count_current = 1
		stored_shells[S.type] -= 1
		caller.to_chat(span("notice","You take a shell from \the [src.name]. There are [stored_shells[S.type]] shells of that type left."))
		if(stored_shells[S.type] <= 0)
			stored_shells -= S.type
		INITIALIZE(S)
		FINALIZE(S)
		I.add_object(S)
		update_sprite()
		return TRUE

	. = ..()
```
perhaps something like this?

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	var/step = max_shells/5
+
+	if(shell_count)
+		var/image/I = new/image(icon,"shell_[CEILING(shell_count/step,1)]")
+		add_overlay(I)
+
+/obj/item/clothing/belt/bandolier/shotgun_12
+	desc_extended = "A not-so-fancy bandolier meant to hold a number of 12 gauge shotgun shells. ALT+Click to grab a handful of shells."
+	icon = 'icons/obj/item/clothing/belts/bandolier_12.dmi'
+	max_shells = 200
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:65]:
```suggestion
/obj/item/clothing/belt/bandolier/shotgun_12
	desc_extended = "A not-so-fancy bandolier meant to hold a number of 12 gauge shotgun shells. ALT+Click to grab a handful of shells."
	icon = 'icons/obj/item/clothing/belts/bandolier_12.dmi'
	max_shells = 200
	
	shell_type = /obj/item/bullet_cartridge/shotgun_12
```
maybe something like this

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		PIERCE = AP_SWORD
+	)
+
+	rarity = RARITY_UNCOMMON
+	value = 200
+
+	var/list/stored_shells = list()
+	var/shell_count = 0
+
+	var/max_shells = 0
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:21]:
```suggestion
/obj/item/clothing/belt/bandolier
	name = "bandolier"
	desc = "For the aspiring sweeper."
	desc_extended = "A not-so-fancy bandolier meant to hold a number of shotgun shells. ALT+Click to grab additional shells while already holding a shell."
	rarity = RARITY_UNCOMMON
	icon = 'icons/obj/item/clothing/belts/bandolier_12.dmi'
	worn_layer = LAYER_MOB_CLOTHING_COAT

	defense_rating = list(
		BLADE = AP_AXE,
		BLUNT = AP_AXE,
		PIERCE = AP_SWORD
	)

	rarity = RARITY_UNCOMMON
	value = 200

	var/list/stored_shells = list()
	var/shell_count = 0

	var/max_shells = 0
	
	var/shell_type = /obj/item/bullet_cartridge
```

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+		caller.to_chat(span("notice","You take a shell from \the [src.name]. There are [stored_shells[S.type]] shells of that type left."))
+		if(stored_shells[S.type] <= 0)
+			stored_shells -= S.type
+		INITIALIZE(S)
+		FINALIZE(S)
+		I.add_object(S)
+		update_sprite()
+		return TRUE
+
+	. = ..()
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:134]:
```suggestion

```

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+/obj/item/clothing/belt/bandolier/shotgun_20
+	desc_extended = "A dyed bandolier meant to hold a number of 20 gauge shotgun shells. ALT+Click to grab a handful of shells."
+	icon = 'icons/obj/item/clothing/belts/bandolier_20.dmi'
+	max_shells = 240
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:139]:
```suggestion
/obj/item/clothing/belt/bandolier/shotgun_20
	desc_extended = "A dyed bandolier meant to hold a number of 20 gauge shotgun shells. ALT+Click to grab a handful of shells."
	icon = 'icons/obj/item/clothing/belts/bandolier_20.dmi'
	max_shells = 240
	
	shell_type = /obj/item/bullet_cartridge/shotgun_20
```

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+		caller.to_chat(span("notice","You take a shell from \the [src.name]. There are [stored_shells[S.type]] shells of that type left."))
+		if(stored_shells[S.type] <= 0)
+			stored_shells -= S.type
+		INITIALIZE(S)
+		FINALIZE(S)
+		I.add_object(S)
+		update_sprite()
+		return TRUE
+
+	. = ..()
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:208]:
```suggestion

```

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+/obj/item/clothing/belt/bandolier/shotgun_23
+	desc_extended = "A surplus bandolier meant to hold a number of 23 gauge shotgun shells. ALT+Click to grab a handful of shells."
+	icon = 'icons/obj/item/clothing/belts/bandolier_23.dmi'
+	max_shells = 160
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:213]:
```suggestion
/obj/item/clothing/belt/bandolier/shotgun_23
	desc_extended = "A surplus bandolier meant to hold a number of 23 gauge shotgun shells. ALT+Click to grab a handful of shells."
	icon = 'icons/obj/item/clothing/belts/bandolier_23.dmi'
	max_shells = 160
	
	shell_type = /obj/item/bullet_cartridge/shotgun_23
```

File: code/_core/obj/item/clothing/belt/bandolier.dm [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+		caller.to_chat(span("notice","You take a shell from \the [src.name]. There are [stored_shells[S.type]] shells of that type left."))
+		if(stored_shells[S.type] <= 0)
+			stored_shells -= S.type
+		INITIALIZE(S)
+		FINALIZE(S)
+		I.add_object(S)
+		update_sprite()
+		return TRUE
+
+	. = ..()
```
Comment (jjpark-kb) [inline code/_core/obj/item/clothing/belt/bandolier.dm:282]:
```suggestion

```

**Comment by Cyaneer** [comment]
I'll see changes proposed and try to work them in.

**Comment by Cyaneer** [comment]
Alright, I swear, last update, now dyeable and working

**Review by BurgerLUA — APPROVED** [review]

**Pull request merged**